### PR TITLE
feat: migrate to server-sent events

### PR DESCRIPTION
**Pull request by jbrunton**

### CODE REVIEW

**Comment by jbrunton-ci-minion** [comment]
#### :tropical_drink: `up` on server-side-events

   <details>
   <summary>Pulumi report</summary>

   ```
   Updating (server-side-events)

View Live: https://app.pulumi.com/jbrunton/chat-demo/server-side-events/updates/5


   pulumi:pulumi:Stack chat-demo-server-side-events running 
   pulumi:pulumi:Stack chat-demo-server-side-events running stack config: {
   pulumi:providers:aws aws  
   pulumi:pulumi:Stack chat-demo-server-side-events running read pulumi:pulumi:StackReference jbrunton/chat-demo-shared-infra/production
   pulumi:pulumi:Stack chat-demo-server-side-events running read pulumi:pulumi:StackReference jbrunton/chat-demo-shared-infra/production
   aws:ecs:Cluster chat-demo-server-side-ev  
   aws:cloudwatch:LogGroup /ecs/chat-demo-server-side-events  
   aws:lb:TargetGroup chat-demo-server-side-ev  
   aws:iam:Role chat-demo-server-side-events  
   aws:lb:ListenerRule chat-demo-server-side-events  
   aws:iam:RolePolicy chat-demo-server-side-events-get-params  
++ aws:ecs:TaskDefinition chat-demo-server-side-events creating replacement [diff: ~containerDefinitions]
   aws:iam:RolePolicyAttachment chat-demo-server-side-events  
++ aws:ecs:TaskDefinition chat-demo-server-side-events created replacement [diff: ~containerDefinitions]
   aws:route53:Record chat-demo-server-side-events  
+- aws:ecs:TaskDefinition chat-demo-server-side-events replacing [diff: ~containerDefinitions]
+- aws:ecs:TaskDefinition chat-demo-server-side-events replaced [diff: ~containerDefinitions]
~  aws:ecs:Service chat-demo-server-side-events updating [diff: ~taskDefinition]
~  aws:ecs:Service chat-demo-server-side-events updated [diff: ~taskDefinition]
-- aws:ecs:TaskDefinition chat-demo-server-side-events deleting original [diff: ~containerDefinitions]
-- aws:ecs:TaskDefinition chat-demo-server-side-events deleted original [diff: ~containerDefinitions]
   pulumi:pulumi:Stack chat-demo-server-side-events  1 message

Diagnostics:
 pulumi:pulumi:Stack (chat-demo-server-side-events):
   stack config: {
    "stackName": "server-side-events",
    "appName": "chat-demo-server-side-events",
    "environment": "development",
    "tag": "HEAD-159bf32",
    "protect": false,
    "domain": "server-side-events.chat-demo.dev.jbrunton-aws.com",
    "publicUrl": "https://server-side-events.chat-demo.dev.jbrunton-aws.com",
    "rootDomain": "jbrunton-aws.com"
   }

Outputs:
   publicUrl: "https://server-side-events.chat-demo.dev.jbrunton-aws.com"

Resources:
   ~ 1 updated
   +-1 replaced
   2 changes. 10 unchanged

Duration: 18s

   ```
   
   </details>

File: server/src/pages/rooms/[id].tsx [hunk 1]
```diff
@@ -63,13 +53,22 @@ const RoomPage: NextPage = () => {
...
+    };
+
+    eventSource.onmessage = (e) => {
+      const message = JSON.parse(e.data);
+      onMessage(message);
+    };
+
+    eventSource.onerror = (e) => {
+      console.log("error", e);
+    };
```
Comment (jbrunton) [inline server/src/pages/rooms/[id].tsx]: this can be removed

File: server/src/pages/rooms/[id].tsx [hunk 2]
```diff
@@ -63,13 +53,22 @@ const RoomPage: NextPage = () => {
...
+
     return () => {
-      socket.close();
+      eventSource?.close();
```
Comment (jbrunton) [inline server/src/pages/rooms/[id].tsx]: This should always be non-null?

File: server/src/app/dependencies/socket-dispatcher.ts [hunk 3]
```diff
@@ -43,26 +41,16 @@ export class SocketDispatcher extends IOServer implements Dispatcher {
...
-      const roomId = socket.handshake.query.roomId as string;
-      socket.join(roomId);
-    });
-  }
-
   private emitPrivateMessage(message: PrivateMessage) {
     logMessage("Sending message (private)", message);
     debug.messages("sending private message: %O", message);
-    this.to(message.recipientId).emit("message", message);
+    this.emitter.emit(message.recipientId, message);
```
Comment (jbrunton) [inline server/src/app/dependencies/socket-dispatcher.ts]: Possibly an opportunity to ensure private messages for a room are restricted to that room. (Currently they appear in all rooms that user is signed into.)

File: server/src/app/dependencies/socket-dispatcher.ts [hunk 4]
```diff
@@ -13,26 +8,29 @@ import {
...
-    });
-    this.configureListeners();
+export class SocketDispatcher implements Dispatcher {
+  private readonly emitter = new EventEmitter();
+
+  constructor(emitter: EventEmitter) {
+    this.emitter = emitter;
+  }
+
+  public subscribe({ userId, roomId, listener }: SubscribeUserParams) {
```
Comment (jbrunton) [inline server/src/app/dependencies/socket-dispatcher.ts]: maybe pass listener as a second param, to make it easier to name the argument something meaningful.

File: server/src/app/dependencies/socket-dispatcher.ts [hunk 5]
```diff
@@ -13,26 +8,29 @@ import {
...
+  listener: (message: Message) => void;
 };
 
-export class SocketDispatcher extends IOServer implements Dispatcher {
-  constructor(httpServer: NetServer) {
-    super(httpServer, {
-      path: "/api/socketio",
-    });
-    this.configureListeners();
+export class SocketDispatcher implements Dispatcher {
```
Comment (jbrunton) [inline server/src/app/dependencies/socket-dispatcher.ts]: This might be better renamed to EventDispatcher or similar

**Pull request merged**